store/unit_store.go

```diff
@@ -89,31 +89,50 @@ func (s unitStores) Refs(f ...RefFilter) ([]*graph.Ref, error) {
 	}
 
 	c_unitStores_Refs_last_numUnitsQueried = 0
-	var allRefs []*graph.Ref
+	var (
+		allRefs   []*graph.Ref
+		allRefsMu sync.Mutex
+	)
+	par := parallel.NewRun(storeFetchPar)
 	for u, us := range uss {
 		if us == nil {
 			continue
 		}
 
-		c_unitStores_Refs_last_numUnitsQueried++
-		setImpliedUnit(f, u)
-		refs, err := us.Refs(filtersForUnit(u, f).([]RefFilter)...)
-		if err != nil && !isStoreNotExist(err) {
-			return nil, err
-		}
-		for _, ref := range refs {
-			ref.UnitType = u.Type
-			ref.Unit = u.Name
-			if ref.DefUnitType == "" {
-				ref.DefUnitType = u.Type
+		par.Do(func() error {
+			// Copy so we can do concurrent modification
+			fCopy := filtersForUnit(u, f).([]RefFilter)
+			setImpliedUnit(fCopy, u)
+
+			refs, err := us.Refs(fCopy...)
+
+			// Use a lock since increment isn't atomic
+			allRefsMu.Lock()
+			c_unitStores_Refs_last_numUnitsQueried++
```

Comment:
this block of code should also be moved to the immediately after `setImpliedUnit...` so that the error check for `refs, err := us.Refs(fCopy...)` can come right after.